### PR TITLE
[Table Designer] api change for insert and move col

### PR DESCRIPTION
**Pull request by caohai**

### CODE REVIEW

File: src/Microsoft.SqlTools.ServiceLayer/TableDesigner/DesignerPathUtils.cs [hunk 1]
```diff
@@ -16,7 +16,7 @@ public static class DesignerPathUtils
         /// Below are the 3 scenarios and their expected path.
         /// Note: 'index-{x}' in the description below are numbers represent the index of the object in the list.
 		/// 1. 'Add' scenario
-		///     a. ['propertyName1']. Example: add a column to the columns property: ['columns'].
+		///     a. ['propertyName1',index-1]. Example: add a column to the columns property: ['columns', 0].
```
Comment (alanrenmsft) [inline src/Microsoft.SqlTools.ServiceLayer/TableDesigner/DesignerPathUtils.cs:19]: so, it is always required to have a index parameter for add, right? if so scenario B should also be updated.

File: src/Microsoft.SqlTools.ServiceLayer/TableDesigner/DesignerPathUtils.cs [hunk 2]
```diff
@@ -16,7 +16,7 @@ public static class DesignerPathUtils
...
 		///     b. ['propertyName1',index-1,'propertyName2']. Example: add a column mapping to the first foreign key: ['foreignKeys',0,'mappings'].
```
Comment (alanrenmsft) [inline src/Microsoft.SqlTools.ServiceLayer/TableDesigner/DesignerPathUtils.cs:20]: we also need to add an index at the end, ADS will also provide an index for ADD operation.

File: src/Microsoft.SqlTools.ServiceLayer/TableDesigner/TableDesignerService.cs [hunk 3]
```diff
@@ -212,12 +215,14 @@ private void HandleAddItemRequest(ProcessTableDesignerEditRequestParams requestP
             var table = this.GetTableDesigner(requestParams.TableInfo).TableViewModel;
             var path = requestParams.TableChangeInfo.Path;
             // Handle the add item request on top level table properties, e.g. Columns, Indexes.
-            if (path.Length == 1)
+            if (path.Length == 2)
             {
                 var propertyName = path[0] as string;
+                var index = path[1] as string;
```
Comment (alanrenmsft) [inline src/Microsoft.SqlTools.ServiceLayer/TableDesigner/TableDesignerService.cs:221]: Convert.ToInt32(path[1]);

File: src/Microsoft.SqlTools.ServiceLayer/TableDesigner/DesignerPathUtils.cs [hunk 4]
```diff
@@ -14,17 +14,19 @@ public static class DesignerPathUtils
...
+		///     b. ['propertyName1',index1,'propertyName2']. Example: update the name of a column: ['columns',0,'name'].
+		///     c. ['propertyName1',index1,'propertyName2',index2,'propertyName3']. Example: update the source column of an entry in a foreign key's column mapping table: ['foreignKeys',0,'mappings',0,'source'].
 		/// 3. 'Remove' scenario
-        ///     a. ['propertyName1',index-1]. Example: remove a column from the columns property: ['columns',0'].
-		///     b. ['propertyName1',index-1,'propertyName2',index-2]. Example: remove a column mapping from a foreign key's column mapping table: ['foreignKeys',0,'mappings',0].
+        ///     a. ['propertyName1',index1]. Example: remove a column from the columns property: ['columns',0'].
+		///     b. ['propertyName1',index1,'propertyName2',index2]. Example: remove a column mapping from a foreign key's column mapping table: ['foreignKeys',0,'mappings',0].
+		/// 4. 'Move' scenario
+        ///     a. ['propertyName1',from,to]. Example: move the second column to the third place: ['columns',1,2].
```
Comment (alanrenmsft) [inline src/Microsoft.SqlTools.ServiceLayer/TableDesigner/DesignerPathUtils.cs:29]: just realized I missed this one in my first review, here the path should be the same as remove:
'propertyName1',index1
'propertyName1',index1,'propertyName2',index2

and the value will be the to Index.

Comment (alanrenmsft) [reply src/Microsoft.SqlTools.ServiceLayer/TableDesigner/DesignerPathUtils.cs:29]: image

File: src/Microsoft.SqlTools.ServiceLayer/TableDesigner/DesignerPathUtils.cs [hunk 5]
```diff
@@ -14,17 +14,20 @@ public static class DesignerPathUtils
         ///<summary>
         /// validate the path in the table designer change information.
         /// Below are the 3 scenarios and their expected path.
-        /// Note: 'index-{x}' in the description below are numbers represent the index of the object in the list.
+        /// Note: 'index{x}' in the description below are numbers represent the index of the object in the list.
 		/// 1. 'Add' scenario
-		///     a. ['propertyName1']. Example: add a column to the columns property: ['columns'].
-		///     b. ['propertyName1',index-1,'propertyName2']. Example: add a column mapping to the first foreign key: ['foreignKeys',0,'mappings'].
+		///     a. ['propertyName1',index1]. Example: add a column to the columns property: ['columns',0].
+		///     b. ['propertyName1',index1,'propertyName2','index2']. Example: add a column mapping to the first foreign key: ['foreignKeys',0,'mappings',0].
```
Comment (alanrenmsft) [inline src/Microsoft.SqlTools.ServiceLayer/TableDesigner/DesignerPathUtils.cs:20]: not quote for index2

File: src/Microsoft.SqlTools.ServiceLayer/TableDesigner/TableDesignerService.cs [hunk 6]
```diff
@@ -691,6 +697,36 @@ private bool HandleUpdateItemRequest(ProcessTableDesignerEditRequestParams reque
...
+                switch (propertyName)
+                {
+                    case TablePropertyNames.Columns:
+                        // TODO add the actual api call
+                        break;
+                    default:
+                        break;
+                }
+            }
+            else if (path.Length == 4)
```
Comment (alanrenmsft) [inline src/Microsoft.SqlTools.ServiceLayer/TableDesigner/TableDesignerService.cs:719]: you don't have to handle this now. we don't need it.

File: src/Microsoft.SqlTools.ServiceLayer/TableDesigner/Contracts/Components/TableProperties.cs [hunk 7]
```diff
@@ -41,6 +41,11 @@ public class TableComponentProperties<T> : ComponentPropertiesBase
         /// </summary>
         public bool CanRemoveRows { get; set; } = true;
 
+        /// <summary>
+        /// Whether rows can be moved.
+        /// </summary>
+        public bool CanMoveRows { get; set; } = true;
```
Comment (alanrenmsft) [inline src/Microsoft.SqlTools.ServiceLayer/TableDesigner/Contracts/Components/TableProperties.cs:47]: I'd set the default value to false.

**Review by alanrenmsft — APPROVED** [review]
Looks good. please remember to test other dacfx features

**Comment by caohai** [comment]
did sanity testing for dacpac, bacpac and schema compare features. things look good. cc @kisantia

**Review by alanrenmsft — APPROVED** [review]

**Pull request merged**